toolkit/tools/internal/retry/retry.go

```diff
@@ -4,13 +4,39 @@
 package retry
 
 import (
+	"math"
 	"time"
 )
 
+// calculateDelay calculates the delay for the given failure count, sleep duration, and backoff exponent base.
+// If the base is negative, it will calculate a linear backoff.
+// If the base is positive, it will calculate an exponential backoff.
+func calculateDelay(failCount int, sleep time.Duration, backoffExponentBase float64) time.Duration {
+	if failCount == 0 {
+		return 0
+	}
+	// Treat negative values of base as linear backoff.
+	if backoffExponentBase < 0 {
+		return sleep * time.Duration(failCount)
+	}
+	// Calculate an exponential backoff.
+	// The formula is: sleep * (backoffExpBase ^ failCount)
+	// For example, if sleep = 1 second, backoffExp = 2.0, failCount = 3
+	// then the delay will be 1 * (2 ^ 3) = 8 seconds.
+	expRetry := math.Pow(backoffExponentBase, float64(failCount-1))
```

Comment:
The `failCount - 1` is inconsistent with the comment above. Let's get them in sync.